### ISSUE TITLE
TestCase not found with BrowserTestBaseDefaultThemeRule

### ISSUE THREAD

**Issue by codebymikey** [mglaman/phpstan-drupal#215]
The `BrowserTestBaseDefaultThemeRule` class throws the false positive error:
```
Class PHPUnit\Framework\TestCase not found.
```
when scanning a normal subclass e.g. Forms, even though it doesn't necessarily depend on PHPUnit itself.

Related to #167

**Comment by codebymikey** [comment]
That makes a lot of sense actually, I've added a similar check into the Rule so that it only does the check when PHPUnit is actually available in the class loader. The alternative is to add a try catch instead for the `ClassAutoloadingException`.

Otherwise the actual crash occurs from within [PHPStan's reflection class](https://github.com/phpstan/phpstan-src/blob/629ccf6b02146e1cbf0499982358c595e3d6edd5/src/Reflection/Runtime/RuntimeReflectionProvider.php#L196):

```
#0 (): PHPStan\Reflection\Runtime\RuntimeReflectionProvider->PHPStan\Reflection\Runtime\{closure}('PHPUnit\Framework\TestCase')
#1 /app/application/web/core/tests/Drupal/Tests/BrowserTestBase.php(40): spl_autoload_call('PHPUnit\Framework\TestCase')
#2 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/vendor/composer/ClassLoader.php(480): include('/app/application/web/core/tests/Drupal/Tests/BrowserTestBase.php')
#3 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/vendor/composer/ClassLoader.php(346): Composer\Autoload\includeFile('/app/application/web/core/tests/Drupal/Tests/BrowserTestBase.php')
#4 (): Composer\Autoload\ClassLoader->loadClass('Drupal\Tests\BrowserTestBase')
#5 (): spl_autoload_call('Drupal\Tests\BrowserTestBase')
#6 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/Runtime/RuntimeReflectionProvider.php(147): class_exists('Drupal\Tests\BrowserTestBase')
#7 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/ClassBlacklistReflectionProvider.php(40): PHPStan\Reflection\Runtime\RuntimeReflectionProvider->hasClass('Drupal\Tests\BrowserTestBase')
#8 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/ChainReflectionProvider.php(25): PHPStan\Reflection\ReflectionProvider\ClassBlacklistReflectionProvider->hasClass('Drupal\Tests\BrowserTestBase')
#9 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/MemoizingReflectionProvider.php(31): PHPStan\Reflection\ReflectionProvider\ChainReflectionProvider->hasClass('Drupal\Tests\BrowserTestBase')
#10 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Broker/Broker.php(54): PHPStan\Reflection\ReflectionProvider\MemoizingReflectionProvider->hasClass('Drupal\Tests\BrowserTestBase')
#11 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Type/ObjectType.php(771): PHPStan\Broker\Broker->hasClass('Drupal\Tests\BrowserTestBase')
#12 /app/application/vendor/mglaman/phpstan-drupal/src/Rules/Drupal/Tests/BrowserTestBaseDefaultThemeRule.php(35): PHPStan\Type\ObjectType->getAncestorWithClassName('Drupal\Tests\BrowserTestBase')
#13 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(69): PHPStan\Rules\Drupal\Tests\BrowserTestBaseDefaultThemeRule->processNode(PhpParser\Node\Stmt\Class_, PHPStan\Analyser\MutatingScope)
#14 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(317): PHPStan\Analyser\FileAnalyser->PHPStan\Analyser\{closure}(PhpParser\Node\Stmt\Class_, PHPStan\Analyser\MutatingScope)
#15 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(259): PHPStan\Analyser\NodeScopeResolver->processStmtNode(PhpParser\Node\Stmt\Class_, PHPStan\Analyser\MutatingScope, Closure)
#16 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(474): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(PhpParser\Node\Stmt\Namespace_, Array, PHPStan\Analyser\MutatingScope, Closure)
#17 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(227): PHPStan\Analyser\NodeScopeResolver->processStmtNode(PhpParser\Node\Stmt\Namespace_, PHPStan\Analyser\MutatingScope, Closure)
#18 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(149): PHPStan\Analyser\NodeScopeResolver->processNodes(Array, PHPStan\Analyser\MutatingScope, Closure)
#19 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/Analyser.php(52): PHPStan\Analyser\FileAnalyser->analyseFile('/app/application/web/modules/contrib/token/src/Controller/TokenAutocompleteController.php', Array, PHPStan\Rules\Registry, NULL)
#20 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Command/AnalyserRunner.php(56): PHPStan\Analyser\Analyser->analyse(Array, NULL, Closure, false, Array)
#21 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Command/AnalyseApplication.php(148): PHPStan\Command\AnalyserRunner->runAnalyser(Array, Array, NULL, Closure, false, true, '/app/application/deprecation_testing.neon', NULL, NULL, _PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput)
#22 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Command/AnalyseApplication.php(79): PHPStan\Command\AnalyseApplication->runAnalyser(Array, Array, false, '/app/application/deprecation_testing.neon', PHPStan\Command\Symfony\SymfonyOutput, PHPStan\Command\Symfony\SymfonyOutput, _PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput)
#23 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/src/Command/AnalyseCommand.php(157): PHPStan\Command\AnalyseApplication->analyse(Array, true, PHPStan\Command\Symfony\SymfonyOutput, PHPStan\Command\Symfony\SymfonyOutput, false, false, '/app/application/deprecation_testing.neon', Array, _PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput)
#24 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Command/Command.php(228): PHPStan\Command\AnalyseCommand->execute(_PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput, _PHPStan_76800bfb5\Symfony\Component\Console\Output\ConsoleOutput)
#25 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(856): _PHPStan_76800bfb5\Symfony\Component\Console\Command\Command->run(_PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput, _PHPStan_76800bfb5\Symfony\Component\Console\Output\ConsoleOutput)
#26 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(237): _PHPStan_76800bfb5\Symfony\Component\Console\Application->doRunCommand(PHPStan\Command\AnalyseCommand, _PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput, _PHPStan_76800bfb5\Symfony\Component\Console\Output\ConsoleOutput)
#27 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(138): _PHPStan_76800bfb5\Symfony\Component\Console\Application->doRun(_PHPStan_76800bfb5\Symfony\Component\Console\Input\ArgvInput, _PHPStan_76800bfb5\Symfony\Component\Console\Output\ConsoleOutput)
#28 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(82): _PHPStan_76800bfb5\Symfony\Component\Console\Application->run()
#29 phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(83): _PHPStan_76800bfb5\{closure}()
#30 /app/application/vendor/phpstan/phpstan/phpstan(8): require('phar:///app/application/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan')

```

**Comment by mglaman** [comment]
@codebymikey can you try this again? The autoloader has

```
        // @todo stop requiring the bootstrap.php and just copy what is needed.
        if (interface_exists(\PHPUnit\Framework\Test::class)) {
            require $this->drupalRoot . '/core/tests/bootstrap.php';
        }
```

**Comment by mglaman** [comment]
I'm finally on a project which has this happening. So I can investigate!

**Comment by mglaman** [comment]
Adding phpstan/phpstan-phpunit should fix it, but in my case it did not for some reason.

```
composer require --dev phpstan/phpstan-phpunit

```

**Comment by tobiasbaehr** [comment]
> Adding phpstan/phpstan-phpunit should fix it, but in my case it did not for some reason.
> 
> ```
> composer require --dev phpstan/phpstan-phpunit
> ```

This does not help. But I found a fix. I will open a PR.

**Comment by mglaman** [comment]
This is fixed, now. Make sure you exclude tests from your analysis.


```
	excludePaths:
		-  */tests/src/*/*.php
```
